examples/js/shaders/SAOShader.js

```diff
@@ -0,0 +1,174 @@
+		"uniform float randomSeed;",
+
+		// RGBA depth
+
+		"#include <packing>",
+
+		"vec3 getViewPosition( const in vec2 screenPosition ) {",
+			"float perspectiveDepth = unpackRGBAToDepth( texture2D( tDepth, screenPosition ) );",
+			"float viewZ = perspectiveDepthToViewZ( perspectiveDepth, cameraNear, cameraFar );",
+			"float clipW = cameraProjectionMatrix[2][3] * viewZ + cameraProjectionMatrix[3][3];",
+			"vec4 clipPosition = vec4( ( vec3( screenPosition, perspectiveDepth ) - 0.5 ) * 2.0, clipW );",
+			"clipPosition.xyz *= clipW;", // unproject to homogeneous coordinates
+			"return ( cameraInverseProjectionMatrix * clipPosition ).xyz;",
+		"}",
+
+		"vec3 getViewNormal( const in vec3 viewPosition, const in vec2 screenPosition ) {",
+			"#if NORMAL_TEXTURE == 1",
+				"return -unpackRGBToNormal( texture2D( tNormal, screenPosition ).xyz );",
+			"#else",
+				"return normalize( cross( dFdx( viewPosition ), dFdy( viewPosition ) ) );",
+			"#endif",
+		"}",
+
+		"vec4 getDiffuseColor( const in vec2 screenPosition ) {",
+			"#if DIFFUSE_TEXTURE == 1",
+				"return texture2D( tDiffuse, vUv );",
+			"#else",
+				"return vec4( 1.0 );",
+			"#endif",
+		"}",
+
+	 "float getOcclusion( const in vec3 viewPosition, const in vec3 viewNormal, const in vec3 viewPositionOffset ) {",
+			"vec3 viewDelta = viewPositionOffset - viewPosition;",
+			"float viewDistance = length( viewDelta );",
+			"float scaledScreenDistance = scale * viewDistance / cameraFar;",
+			"return intensity * max(0.0, (dot(viewNormal, viewDelta) - minResolution * cameraFar) / scaledScreenDistance - bias) / (1.0 + pow2( scaledScreenDistance ) );",
+		"}",
+
+		"float getAmbientOcclusion( const in vec3 viewPosition ) {",
+
+			"vec3 viewNormal = getViewNormal( viewPosition, vUv );",
+
+			"float random = rand( vUv + randomSeed );",
+			"vec2 radius = vec2( kernelRadius ) / size;",
+			"float numSamples = float( NUM_SAMPLES );",
+			"float numRings = float( NUM_RINGS );",
+			"float alphaStep = 1.0 / numSamples;",
+
+			// jsfiddle that shows sample pattern: https://jsfiddle.net/a16ff1p7/
+
+			"float occlusionSum = 0.0;",
+			"float alpha = 0.0;",
+			"float weight = 0.0;",
+
+			"for( int i = 0; i < NUM_SAMPLES; i ++ ) {",
+				"float angle = PI2 * ( numRings * alpha + random );",
+				"vec2 currentRadius = radius * ( 0.01 + alpha * 0.99 );",
+				"vec2 offset = vec2( cos(angle), sin(angle) ) * currentRadius;",
+				"alpha += alphaStep;",
+
+				"vec3 viewPositionOffset = getViewPosition( vUv + offset );",
+				"if( -viewPositionOffset.z >= cameraFar ) {",
+					"continue;",
+				"}",
+
+				"occlusionSum += getOcclusion( viewPosition, viewNormal, viewPositionOffset );",
+				"weight += 1.0;",
+
+			"}",
+
+			"if( weight == 0.0 ) return 0.0;",
+			"return occlusionSum / weight;",
+
+		"}",
+
+
+		"void main() {",
+
+			"gl_FragColor = getDiffuseColor( vUv );",
+
+			"vec3 viewPosition = getViewPosition( vUv );",
+			"if( -viewPosition.z >= cameraFar ) {",
+				"return;",
+			"}",
```

Comment:
Oh nice! That one's been worth it, then.

I guess you may save another texture fetch by doing `getDiffuseColor` after the test. Won't be a huge difference, but may not necessarily optimize out either, since writing directly to an output makes it a fairly special case for the compiler to consider.